 ---
# Data Challenge ULACTI 2021
#### Licitaciones del PODER JUDICIAL
---
Estudiante: 

Jeffry Luque

Sofía Ramírez

Luis V

Grupo: 8

## - Análisis de datos de las licitaciones publicas*
>Trabajo realizado para obtener información sobre los procesos de licitación en el Poder Judicial del 2016 al 2020, en busca de mejorar para beneficiar a la institución en sus procesos de compra.
---

## Datos Abiertos Poder Judicial (PJ)

Para proponer una solución o mejoras para los procesos de licitación, compras o adquisiones de servicios en el Poder Judicial se realizará un estudio de los datos suministrados por la misma institución, con respecto a la compra de servicios y bienes en los años 2016,2017,2018,2019 y 2020.

Una vez con los datos, se procederá a realizar comparaciones y análisis del proceso, para veririficar o no el buen desempeño en el proceso.

### Partida

Como primer acercamiento partiremos de los datos del 2016, para luego realizar una comparación con los siguientes años. Se comienza con la importación de las librerías a utilizar en python, se llama los archivos iniciales a analizar, y se proceden a conocer las columnas que nos brindan estos archivos: 

In [75]:
import pandas as pd
import requests

oferente = 0
datos16 = pd.read_excel('2016_Poder_Judicial_CR.xls')
parties16  = pd.read_csv('parties.csv')
print("Columnas de Empresas Oferentes: ")
print(list(datos16.columns))
print("\n Columnas de releases: ")
print(list(parties16.columns))

Columnas de Empresas Oferentes: 
['ocid', 'id', 'date', 'tag', 'initiationType', 'buyer/id', 'buyer/name', 'tender/id', 'tender/title', 'tender/description', 'tender/status', 'tender/procuringEntity/id', 'tender/procuringEntity/name', 'tender/value/amount', 'tender/value/currency', 'tender/procurementMethod', 'tender/mainProcurementCategory', 'tender/awardCriteria', 'tender/submissionMethod', 'tender/tenderPeriod/startDate', 'tender/tenderPeriod/endDate', 'tender/enquiryPeriod/startDate', 'tender/enquiryPeriod/endDate', 'tender/awardPeriod/startDate', 'tender/awardPeriod/endDate', 'language', 'planning/budget/description']

 Columnas de releases: 
['ocid', 'id', 'parties/0/id', 'parties/0/name', 'parties/0/identifier/id', 'parties/0/identifier/legalName', 'parties/0/address/streetAddress', 'parties/0/roles', 'parties/0/contactPoint/email', 'parties/0/contactPoint/telephone', 'parties/0/contactPoint/faxNumber', 'parties/0/contactPoint/name', 'parties/0/address/postalCode']


### Funciones

Se crea una función para enviar los oferentes que se desean analizar de acuerdo al año, con respecto a las actividades generadas en la licitación, con esto se busca encontrar cuantas empresas son oferentes en las distintas actividades que el Poder Judicial abre el proceso de licitación y contratación, se desea analizar si hay una correlación entre pocas empresas oferentes con fechas cortas habilitadas para ofertar, es decir que por ejemplo solo hayan 2 días para ofertar afecte que pocas empresas participen.

Primero vemos como se muestra en los datos del PJ las actividades a realizar:

In [19]:
print(parties16["ocid"])

0       ocds-fnha3a-000001-2016
1       ocds-fnha3a-000001-2016
2       ocds-fnha3a-000001-2016
3       ocds-fnha3a-000001-2016
4       ocds-fnha3a-000001-2016
                 ...           
8936    ocds-fnha3a-003134-2016
8937    ocds-fnha3a-003134-2016
8938    ocds-fnha3a-003135-2016
8939    ocds-fnha3a-003135-2016
8940    ocds-fnha3a-003135-2016
Name: ocid, Length: 8941, dtype: object


Notamos que en el 2016, por ejemplo, se desarrollaron 8941 actividades que contaron con al menos una empresa oferente, los String generados son similares excepto la numeración final, por lo que procedemos con la siguiente función para recibir los datos, y esta devuelva una lista con la cantidad de oferentes por cada actividad:

In [39]:
def cantidad_Oferentes(parties):
    id_Oferente = list(parties["ocid"])
    oferente = 0
    actividad = []
    id_num_of = []
    array =[]
    listaN =[]
    for n in id_Oferente:
        num = int(n[13:18])
        id_num_of.append(num)
    m =1

    data = sorted(id_num_of)
    aux = id_num_of[0] #Se iguala al primer elemento de la lista
    count = 0 #se inicia el contador en 0
    for i in id_num_of:
        if aux == i:
          count += 1
        else:
          array.append(count)
          count = 1
          aux = i
    return array


A continuación se llama la función creada con los oferentes del año a analizar, por ejemplo el 2016, solamente mostrandos las primeras 10 actividades:

In [56]:
oferentes16 = cantidad_Oferentes(parties16)
for n in range(0,11):
    print("Cantidad de oferentes en la actividad número ",n, " son", oferentes[n])

Cantidad de oferentes en la actividad número  0  son 7
Cantidad de oferentes en la actividad número  1  son 19
Cantidad de oferentes en la actividad número  2  son 2
Cantidad de oferentes en la actividad número  3  son 2
Cantidad de oferentes en la actividad número  4  son 3
Cantidad de oferentes en la actividad número  5  son 2
Cantidad de oferentes en la actividad número  6  son 6
Cantidad de oferentes en la actividad número  7  son 3
Cantidad de oferentes en la actividad número  8  son 2
Cantidad de oferentes en la actividad número  9  son 4
Cantidad de oferentes en la actividad número  10  son 11


Contando con las 8941 actividades del 2016, verificaremos cuantas actividades tuvieron 3 o menos empresas oferentes, una vez encontradas se hará una verificación de las fechas en la licitación:

In [60]:
def posiciones_Act(oferentes):
    array2=[]
    act = []
    m, pos= 0,0
    for i in oferentes:
        if(i<=3):
            m+=1
            act.append(pos)
        pos+=1
    return act

A continuación se llama la función creada con las actividades  a analizar que tienen  tres o menos oferentes, por ejemplo el 2016 estas son las primeras 10 actividades que cumplen esos números:

In [67]:
act = posiciones_Act(oferentes16)
for n in range(0,11):
    print("La actividad número ",act[n], " tiene menos de 3 oferentes")

La actividad número  2  tiene menos de 3 oferentes
La actividad número  3  tiene menos de 3 oferentes
La actividad número  4  tiene menos de 3 oferentes
La actividad número  5  tiene menos de 3 oferentes
La actividad número  7  tiene menos de 3 oferentes
La actividad número  8  tiene menos de 3 oferentes
La actividad número  12  tiene menos de 3 oferentes
La actividad número  18  tiene menos de 3 oferentes
La actividad número  19  tiene menos de 3 oferentes
La actividad número  20  tiene menos de 3 oferentes
La actividad número  22  tiene menos de 3 oferentes


### Fechas

Con los datos a analizar, por ejemplo 2016, se analizarán las actividades que cuentan con tres o menos oferentes, se visualizarán las fechas en la que las licitaciones estaban abiertas, es decir, que las empresas podían participar para la venta o contratación.

Primero vemos la tabla general, con las primeras cinco actividades, para conocer los datos proporcionados:

In [68]:
datos16.head(5)

ocid                    id                       date  \
0  ocds-fnha3a-000001-2016    2016LN-000001-PROV  2016-01-07T10:54:00-06:00   
1  ocds-fnha3a-000002-2016  2016CD-000001-PROVCD  2016-01-07T13:16:00-06:00   
2  ocds-fnha3a-000004-2016  2016CD-000003-PROVCD  2016-01-07T15:16:00-06:00   
3  ocds-fnha3a-000005-2016  2016CD-000004-PROVCD  2016-01-11T11:28:00-06:00   
4  ocds-fnha3a-000006-2016  2016CD-000005-PROVCD  2016-01-11T11:42:00-06:00   

      tag initiationType      buyer/id  \
0  tender         tender  2-300-042155   
1  tender         tender  2-300-042155   
2  tender         tender  2-300-042155   
3  tender         tender  2-300-042155   
4  tender         tender  2-300-042155   

                                 buyer/name    tender/id  \
0  Corte Suprema de Justicia Poder Judicial  000001-2016   
1  Corte Suprema de Justicia Poder Judicial  000002-2016   
2  Corte Suprema de Justicia Poder Judicial  000004-2016   
3  Corte Suprema de Justicia Poder Judicial  000005-2016   
4  Corte Suprema de Justicia Poder Judicial  000006-2016   

                                        tender/title  \
0         Compra de LLantas y baterías según demanda   
1  Reparación de canoas, precintas, aleros y pint...   
2  Construcción de aceras, en el costado sur, nor...   
3  Compra de UPS para centros de datos de Osa y C...   
4  Remodelación de servicios sanitarios de emplea...   

                                  tender/description  ...  \
0         Compra de LLantas y baterías según demanda  ...   
1  Reparación de canoas, precintas, aleros y pint...  ...   
2  Construcción de aceras, en el costado sur, nor...  ...   
3  Compra de UPS para centros de datos de Osa y C...  ...   
4  Remodelación de servicios sanitarios de emplea...  ...   

  tender/awardCriteria tender/submissionMethod tender/tenderPeriod/startDate  \
0        ratedCriteria                inPerson     2016-05-06T00:00:00-06:00   
1        ratedCriteria                inPerson     1900-01-01T00:00:00-06:00   
2        ratedCriteria                inPerson     2016-03-15T15:26:43-06:00   
3        ratedCriteria                inPerson     2016-01-27T15:14:23-06:00   
4        ratedCriteria                inPerson     2016-01-18T00:00:00-06:00   

   tender/tenderPeriod/endDate tender/enquiryPeriod/startDate  \
0    2016-06-07T10:00:00-06:00      2016-05-06T00:00:00-06:00   
1    1900-01-01T00:00:00-06:00      1900-01-01T00:00:00-06:00   
2    2016-03-29T10:00:00-06:00      2016-03-15T15:26:43-06:00   
3    2016-02-02T10:00:00-06:00      2016-01-27T15:14:23-06:00   
4    2016-01-25T09:30:00-06:00      2016-01-18T00:00:00-06:00   

  tender/enquiryPeriod/endDate tender/awardPeriod/startDate  \
0    2016-06-07T10:00:00-06:00    2016-06-07T10:00:00-06:00   
1    1900-01-01T00:00:00-06:00    1900-01-01T00:00:00-06:00   
2    2016-03-29T10:00:00-06:00    2016-03-29T10:00:00-06:00   
3    2016-02-02T10:00:00-06:00    2016-02-02T10:00:00-06:00   
4    2016-01-25T09:30:00-06:00    2016-01-25T09:30:00-06:00   

  tender/awardPeriod/endDate language planning/budget/description  
0  2016-08-10T00:00:00-06:00       es                         NaN  
1  1900-01-01T00:00:00-06:00       es                         NaN  
2  2016-04-12T00:00:00-06:00       es     Múltiples requisiciones  
3  2016-02-16T00:00:00-06:00       es     Múltiples requisiciones  
4  2016-02-08T00:00:00-06:00       es                         NaN  

[5 rows x 27 columns]

Se empezará analiazando  el "ocid" que son los id de las actividades específicamente con  las fechas de licitación "tender/tenderPeriod". 

In [83]:
id_Oferente = list(datos16["ocid"])
oferente = 0
actividad = []
id_num_of = []
array =[]
listaN =[]
for n in id_Oferente:
    num = int(n[13:18])
    id_num_of.append(num)


In [117]:
fechas = datos16["tender/awardPeriod/startDate"]
m =0
for n in act:
    for m in range(0, len(act)):
        if int(datos16["ocid"][m][13:18]) == n:
            print(fechas[m])

1900-01-01T00:00:00-06:00
2016-03-29T10:00:00-06:00
2016-02-02T10:00:00-06:00
2016-02-03T09:30:00-06:00
2016-01-12T10:00:00-06:00
2016-02-16T10:00:00-06:00
1900-01-01T00:00:00-06:00
2016-04-22T10:00:00-06:00
2016-06-27T10:00:00-06:00
2016-01-20T10:00:00-06:00
2016-01-26T09:30:00-06:00
2016-01-27T09:30:00-06:00
2016-02-05T09:30:00-06:00
2016-03-18T10:00:00-06:00
2016-01-21T13:30:00-06:00
2016-01-22T15:30:00-06:00
2016-02-01T10:00:00-06:00
2016-01-26T14:00:00-06:00
2016-04-08T09:30:00-06:00
2016-02-02T14:00:00-06:00
2016-01-26T12:00:00-06:00
2016-02-11T09:00:00-06:00
2016-02-02T14:00:00-06:00
2016-02-03T10:00:00-06:00
2016-02-02T15:00:00-06:00
2016-02-03T09:30:00-06:00
2016-01-29T10:00:00-06:00
2016-01-27T12:00:00-06:00
2016-01-29T15:00:00-06:00
2016-01-27T15:00:00-06:00
2016-01-27T16:00:00-06:00
2016-02-01T15:00:00-06:00
2016-01-28T09:00:00-06:00
2016-02-05T15:00:00-06:00
2016-02-01T10:00:00-06:00
2016-03-09T09:30:00-06:00
2016-02-05T14:00:00-06:00
2016-03-18T09:30:00-06:00
2016-03-09T1

In [82]:
from datetime import datetime, timedelta

for n in act:
    fecha_cad1 = '2016-01-15'
    fecha_cad2 = '2016-01-18'
    fecha1 = datetime.strptime(fecha_cad1, '%Y-%m-%d')
    fecha2 = datetime.strptime(fecha_cad2, '%Y-%m-%d')
    dias = (fecha2 - fecha1) / timedelta(days=1)
    print(dias)

3.0
